File: 2369-check-if-there-is-a-valid-partition-for-the-array/2369-check-if-there-is-a-valid-partition-for-the-array.py
class Solution:
    def dp(self, i: int, memo: Dict[int, int], nums: List[int]) -> bool:
        if i in memo:
            return memo[i]
        ans = False
        
        if i > 0 and nums[i] == nums[i - 1]:
            ans |= self.dp(i - 2, memo, nums)
        if i > 1 and nums[i] == nums[i - 1] == nums[i - 2]:
            ans |= self.dp(i - 3, memo, nums)
        if i > 1 and nums[i] == nums[i - 1] + 1 == nums[i - 2] + 2:
            ans |= self.dp(i - 3, memo, nums)
        memo[i] = ans
        return ans
    
    def validPartition(self, nums: List[int]) -> bool:
        n, memo = len(nums), {-1: True}
        return self.dp(n - 1, memo, nums)

File: 2369-check-if-there-is-a-valid-partition-for-the-array/2369-check-if-there-is-a-valid-partition-for-the-array.cpp
class Solution {
public:
    bool validPartition(vector<int>& nums) {
        int n = nums.size();
        vector<bool> dp(n + 1);
        dp[0] = true;
        
        for (int i = 0; i < n; ++i){
            int dpIndex = i + 1;
            if (i > 0 && nums[i] == nums[i - 1]) dp[dpIndex] = dp[dpIndex] | dp[dpIndex - 2];
            if (i > 1 && nums[i] == nums[i - 1] && nums[i - 1] == nums[i - 2]) dp[dpIndex] = dp[dpIndex] | dp[dpIndex - 3];
            if (i > 1 && nums[i] == nums[i - 1] + 1 && nums[i - 1] + 1 == nums[i - 2] + 2) dp[dpIndex] = dp[dpIndex] | dp[dpIndex - 3];
        }
        
        return dp[n];
    }
};